columnflow/tasks/reduction.py

```diff
@@ -564,54 +564,75 @@ def _req_merged_reduced_events(self, **params) -> law.Task:
+        # - when merging is forced, require it
+        # - otherwise, and if the merging is already known, require either reduced or merged events
+        if self.skip_merging or (not self.force_merging and self.dataset_info_inst.n_files == 1):
+            # reduced events are used directly without having to look into the file merging factor
             if not self.pilot:
                 reqs["events"] = self._req_reduced_events()
         else:
-            req_merged = self._req_merged_reduced_events()
-            if self.force_merging or req_merged.complete():
-                reqs["events"] = req_merged
+            # here, the merging is unclear so require the stats
+            reqs["reduction_stats"] = self.reqs.MergeReductionStats.req(self)
+
+            if self.force_merging:
+                # require merged events when forced
+                reqs["events"] = self._req_merged_reduced_events()
+            else:
+                # require either when the file merging is known, and nothing otherwise to let the
+                # dynamic dependency definition resolve it at runtime
+                file_merging = self.file_merging
+                if file_merging > 1:
+                    reqs["events"] = self._req_merged_reduced_events()
+                elif file_merging == 1 and not self.pilot:
+                    reqs["events"] = self._req_reduced_events()
 
         return reqs
 
     def requires(self):
-        # same as for workflow requirements, declare static requirements when a decision is pre-set,
-        # and otherwise let run() yield dynamic ones
-        if self.skip_merging:
-            return self._req_reduced_events()
-        req_merged = self._req_merged_reduced_events()
-        if self.force_merging or req_merged.complete():
-            return req_merged
-        return []
+        # same as for workflow requirements without optional pilot check
+        reqs = {}
+        if self.skip_merging or (not self.force_merging and self.dataset_info_inst.n_files == 1):
+            reqs["events"] = self._req_reduced_events()
+        else:
+            reqs["reduction_stats"] = self.reqs.MergeReductionStats.req(self)
+
+            if self.force_merging:
+                reqs["events"] = self._req_merged_reduced_events()
+            else:
+                file_merging = self.file_merging
+                if file_merging > 1:
+                    reqs["events"] = self._req_merged_reduced_events()
+                elif file_merging == 1:
+                    reqs["events"] = self._req_reduced_events()
+
+        return reqs
 
     @workflow_condition.output
     def output(self):
-        if self.skip_merging or self.force_merging:
-            req = self.requires()
+        # determine the definitive requirement
+        if (
+            self.skip_merging or
+            self.dataset_info_inst.n_files == 1 or
+            (not self.force_merging and self.file_merging == 1)
```

Comment:
When `self.force_merging` is True, we should require the merged reduced events even if the number of files is 1, right?

```suggestion
            self.skip_merging or
            (not self.force_merging and (self.dataset_info_inst.n_files == 1 or self.file_merging == 1))
```

Reply:
Good point. And I guess this can be done even easier by just using `self.requires()["events"]` which is definitely known at this point. I did that in ab9c1ac9.

Reply:
true, this is much better :+1: